File: Models/Composer.cs
using System;
using JetBrains.Annotations;

namespace Models
{
    /// <summary>
    /// Composer, like Bach or Beethoven.
    /// </summary>
    [PublicAPI]
    public record Composer
    {
        /// <summary>
        /// Id in the database.
        /// </summary>
        public int Id { get; init; }

        /// <summary>
        /// First name.
        /// </summary>
        /// <example>Ludwig van</example>
        public string FirstName { get; init; } = "";

        /// <summary>
        /// Last name.
        /// </summary>
        /// <example>Beethoven</example>
        public string LastName { get; init; } = "";
        
        /// <summary>
        /// Year composer was born. 
        /// </summary>
        public int YearBorn { get; init; }
        
        /// <summary>
        /// Year composer died, can be null for alive composers.
        /// </summary>
        public int? YearDied { get; init; }

        /// <summary>
        /// Countries composer is associated with.
        /// </summary>
        public string[] Countries { get; init; } = Array.Empty<string>();

        /// <summary>
        /// Unique composer readable text id, to be used in URLs.
        /// </summary>
        /// <example>beethoven</example>
        public string Slug { get; init; } = "";
        
        /// <summary>
        /// Link to composer's wikipedia page.
        /// </summary>
        public string? WikipediaLink { get; init; }
        
        /// <summary>
        /// Link to composer's IMSLP page with the list of all works.
        /// </summary>
        public string? ImslpLink { get; init; }
        
        /// <summary>
        /// Show this composer on the main page
        /// </summary>
        public bool Enabled { get; init; }
    }
}

File: Site/Domain/Composer.fs
/// Business logic for Composer.
module Site.Domain.Composer

open FSharp.Json
open Giraffe
open Microsoft.AspNetCore.Http
open Site.Helpers
open Site.Postgres
open Site.Redis

/// Composer, like Bach or Beethoven
type Composer =
    { id: int
      firstName: string
      lastName: string
      yearBorn: int
      yearDied: int option
      countries: string list // Countries composer is associated with.
      slug: string // Unique composer readable text id, to be used in URLs.
      wikipediaLink: string option
      imslpLink: string option
      enabled: bool } // Show this composer on the main page

/// Returns composer by URL slug.
let getComposer (slug: string) : Composer option =
    let redisKey = "opusclassical:composer:" + slug

    match retrieveRedis redisKey with
    | Some c -> Json.deserialize<Composer> c |> Some
    | None ->
        let sql = "select composer_by_slug(@ComposerSlug) as json"
        let parameters = [ "ComposerSlug", Sql.text slug ] |> Some

        let json =
            query (sql, parameters, jsonMapper) |> Async.RunSynchronously

        storeRedis (redisKey, json.Head, expire.Long) |> ignore
        json.Head |> Json.deserialize<Composer> |> Some
